cmd/root.go

```diff
@@ -44,7 +44,18 @@ to quickly create a Cobra application.`,
 		}
 
 		for _, v := range files {
-			id, err := dox.Publish(v, root, dryRun)
+			_, err := dox.Publish(v, root, repoRoot, dryRun)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "error: %s\n", err)
+				os.Exit(1)
+			}
+		}
+
+		// ham-fisted approach to ensuring that all relative links are linked
```

Comment:
I really don't like this, but without fundamental changes to the way dox does its publishing, I can't think of a workable solution right now.

Reply:
Pros:
* Works
* ???
Cons:
* Publishes twice
* ???

Would you add anything to either of these lists? Is the primary reason for double publishing because there are no guarantees about the order of publishing, and potentially also recursive links (page A links to page B, page B links to page A)? If so, how would you go about improving this situation?

Reply:
I'd say the primary reason is that we need to know the page IDs for the dox source files in order to relatively link to them. Same problem can be seen in [image support ](https://github.com/jesselang/dox/blob/master/pkg/publish.go#L92-L103) where when a page is first created, we can't attach files to it at the same time since the attachment endpoint needs a page ID.

This is a little different because a dox source file could link to another dox source file that hasn't been published yet, so the relative link would not be replaced. This way to guarantee that all dox source files have page IDs is to publish them all, then publish again for relative linking. I don't like it.

One alternative might be to perform some sort of pre-pass, where dox ensures that all dox source files are published (could be existing so no work, or new dox source so create a blank page) and in the main pass proceed to "update" all pages. This would also help with image support.

Reply:
Latest commit has code to address this problem.